subsys/cpp/cpp_init.c

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2021 Synopsys, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Author: Evgeniy Paltsev
+ */
+
+#ifdef CONFIG_CPP_STATIC_INIT_GNU
+
+void __do_global_ctors_aux(void);
+void __do_init_array_aux(void);
+
+void z_cpp_init_static(void)
+{
+	__do_global_ctors_aux();
+	__do_init_array_aux();
+}
+
+#else
```

Comment:
This _init is from MetaWare, right? Shouldn't it be #if'def-ed for MetaWare?
Besides C++ constructors, are there other side effects of MetaWare _init()? If so, would there be lower-level MetaWare functions that only initialize the constructors?

Reply:
The _init is from MetaWare and actually it's #if'def-ed for MetaWare with `CONFIG_CPP_STATIC_INIT_GNU` option which is disabled for MetaWare toolchain.

_init is actually lower-level and only calls constructors of static objects (yes, the naming is misleading a bit)

Reply:
ok, so _init is only the constructors, good.
My point about #ifdef for MetaWare is that CONFIG_CPP_STATIC_INIT_GNU is very generic: not GNU can be MetaWare, but also ARM CLANG, or other 3rd party toolchains. Hence my suggestion to make this MetaWare specific, not all "non-GNU".

Reply:
Now all the MWDT-related code is guarded with appropriate if'def-s. I've kept the option name (`CONFIG_CPP_STATIC_INIT_GNU`) but now it can be used for other toolchains if it is required.

Reply:
Thanks - it's the #ifdef __CCAC__ that I missed to indicate the _init() is something metaware specific.